File: cpp/leetcode/PairDistance.cpp
//Leetcode Problem No 719 Find K-th Smallest Pair Distance
//Solution written by Xuqiang Fang on 19 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        const int n = nums.size();
        sort(nums.begin(), nums.end());
        int l = 0;
        int r = nums[n-1];
        while(l <= r){
            int m = (l + r) / 2;
            int count = 0;
            int dis = 0;
            int j = n-1;
            for(int i=n-2; i>=0; --i){
                while(j > i && nums[j] - nums[i] > m){
                    --j;
                }
                count += j - i;
                if(nums[j] - nums[i] > dis){
                    dis = nums[j] - nums[i];
                }
            }
            if(count == k || l == r){
                return dis;
            }
            else if(count > k){
                r = m;
            }
            else{
                l = m + 1;
            }
        }
        return 0;
    }
};

int main(){
    Solution s;
    vector<int> nums{1,3,1};
    vector<int> num{9,10,7,10,6,1,5,4,9,8};
    cout << s.smallestDistancePair(nums, 1) << endl;
    cout << s.smallestDistancePair(nums, 2) << endl;
    cout << s.smallestDistancePair(nums, 3) << endl;
    cout << s.smallestDistancePair(num, 18) << endl;
    cout << s.smallestDistancePair(num, 19) << endl;
    return 0;
}

File: python/leetcode/PairDistance.py
'''
Leetcode problem No 719 Find K-th Smallest Pair Distance
Solution written by Xuqiang Fang on 19 June, 2018
'''

class Solution(object):
    def smallestDistancePair(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        nums.sort()
        print(nums)
        n = len(nums)
        l = 0; r = nums[n-1]
        while(l <= r):
            m = (l + r) // 2
            count = 0; dis = 0; j = n-1
            for i in range(n-2, -1, -1):
                while j > i and nums[j]-nums[i] > m:
                    j -= 1
                count += j-i
                if nums[j]-nums[i] > dis:
                    dis = nums[j] - nums[i]
            if count == k or l == r:
                return dis
            elif count > k:
                r = m
            else:
                l = m+1
        return 0
        

def main():
    s = Solution()
    nums = [1,3,1]
    num = [9,10,7,10,6,1,5,4,9,8]
    print(s.smallestDistancePair(nums, 1))
    print(s.smallestDistancePair(nums, 2))
    print(s.smallestDistancePair(nums, 3))
    print(s.smallestDistancePair(num, 32))
main()
